### ISSUE TITLE
Reading time not changing with "setBusVoltageConversionTime"

### ISSUE THREAD

**Issue by joachymus** [RobTillaart/INA3221_RT#8]
Hi all,

I am just playing with INA3221 and this amazing library. Unfortunately, I am not sure whether the setting functions (average, busConvTine, ShuntConvTime) work properly/at all. I cannot tell from the values but there is definitely no "time spent" change. I do all readings within a milli second regardless what values are set to those parametres...

With all zeros:
```cpp
INA.setAverage(0);
INA.setBusVoltageConversionTime(0);
INA.setShuntVoltageConversionTime(0);
```
the reading is done within the same time as with all sevens:
```cpp
INA.setAverage(7);
INA.setBusVoltageConversionTime(7);
INA.setShuntVoltageConversionTime(7);
```

I am not an expert but if feels improper... Am I misunderstanding anything here? Please advise.

Thanks in advance and have a great day.

Br,
j.

**Comment by RobTillaart** [comment]
Thanks for reporting the issue, 
At the moment I have more than a dozen issues open and try to work through it.
When time permits I will try to have a first look this weekend.

Rob

**Comment by RobTillaart** [comment]
@joachymus 
Sunday afternoon, got some time

You found a bug, 👍
```cpp
int INA3221::setAverage(uint8_t avg)
{
  if (avg > 7) return -10;
  uint16_t mask = _readRegister(INA3221_CONFIGURATION);
  mask &= ~(7 << 9);
  mask |= (avg << 9);  <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
  //  in one line mask &= ~((~avg) << 9)
  return _writeRegister(INA3221_CONFIGURATION, mask);
}
```

as avg is an 8 bit variable a shift over 9 places will effectively make it zero.

Will look into all of them, and create a develop branch asap.

**Comment by RobTillaart** [comment]
To fix
- [x] INA.setAverage(0);
- [x] INA.setBusVoltageConversionTime(0);
- [x] INA.setShuntVoltageConversionTime(0);

**Comment by RobTillaart** [comment]
@joachymus 

Created a develop branch with fixes, if you have time can you please verify if they work?

**Comment by RobTillaart** [comment]
@joachymus 

Also created a PR as preparation for the 0.4.0 version.

Did you encounter other missing or failing functionality? 
As stated in the readme.md I did not test all (sorry, time is scarce).

**Comment by joachymus** [comment]
Hi Rob,

Thank you very much for getting back to me and for addressing my concern. It took me some time to unriddle the issue/pull request logic here but I finally got your message :)
I already shut off my workshop fo rtoday but I will definitely look into it tomorrow.
To be honest, I am not exactly a master in C++ (in any programming language for that matter) and in bitwise register shifting and masking I am most probably reaching my brain capacity...
Anyway, I will try and let you know.

Once again, thank you very much and have a great evening.
BR,
j.

**Comment by RobTillaart** [comment]
> thank you very much and have a great evening.

You're welcome!

**Comment by joachymus** [comment]
Hi Rob,

I implemented all changes, unfortunatelly no success :(. Anyway, it looks like the register is being set properly, see below.
This is my code:
```cpp
  bus_voltage = 0;
  shunt_voltage = 0;
  for (int i = 1; i <= 10; i++) {                                              - I wanted one extra decimal for "precision" (don't ask...)
    bus_voltage += INA.getBusVoltage(channel);
    shunt_voltage += INA.getShuntVoltage(channel);
  Serial.println(millis());                                                        - this is where I get the milli seconds                     
  }
  Serial.println(INA.getConfiguration());                              - When reading the config I get all params HIGH (0100111111111111) 
  bus_voltage /= 10.0;
  shunt_voltage /= 10.0;

  // bus_voltage = INA.getBusVoltage(channel);                - My original readings; did not work neither
  // shunt_voltage = INA.getShuntVoltage(channel);
```

Serial Output - millis plus config (readingg done each 300ms):
```
68100, 68102, 68103, 68104, 68105, 68106, 68107, 68108, 68109, 68110
20479
68400, 68402, 68403, 68404, 68405, 68406, 68407, 68408, 68409, 68410
20479
```

I cannot promis that the problem is not elswhere in my code - anyway, the combination "all set high" and still getting quick readings makes me confused...

Any other idea?
I don't want to take your precious time so let it be if not interested - this is not life threatening for my hobby project :) !!

Anyway, thanks again for your support and an amazing contribution in general!
BR,
j.

**Comment by RobTillaart** [comment]
(updated your post for syntax highlighting / easier read, have a look how its done)

Every problem needs time to solve, it still might be a problem in the library. 

Can you run this program. It tests the timing of average.  (will add it to the repo asap)
```cpp
//
//    FILE: INA3221_test_average.ino
//  AUTHOR: Rob Tillaart
// PURPOSE: test averaging single channel
//     URL: https://github.com/RobTillaart/INA3221_RT


#include "INA3221.h"


INA3221 INA(0x40);
uint32_t start, stop;


void setup()
{
  Serial.begin(115200);
  Serial.println(__FILE__);
  Serial.print("INA3221_LIB_VERSION: ");
  Serial.println(INA3221_LIB_VERSION);
  Serial.println();

  Wire.begin();

  if (!INA.begin() )
  {
    Serial.println("could not connect. Fix and Reboot");
    while(1);
  }
  else
  {
    Serial.print("Found: \t");
    Serial.println(INA.getAddress());
  }

  //  overwrite default shunts.
  INA.setShuntR(0, 0.100);


  delay(100);  // fllussssh IO

  for (int avg = 0; avg < 8; avg++)
  {
    INA.setAverage(avg);
    Serial.print("AVG:\t");
    Serial.println(INA.getAverage());
    delay(100);

    start = micros();
    float busVoltage = INA.getBusVoltage(0);
    stop = micros();
    Serial.print("BUS V:\t");
    Serial.print(busVoltage, 3);
    Serial.print("\t");
    Serial.println(stop - start);
    delay(100);
  }
}


void loop()
{
}


//  -- END OF FILE --
```

**Comment by joachymus** [comment]
Output:
AVG:	1
BUS V:	0.000	492
AVG:	2
BUS V:	0.000	492
AVG:	3
BUS V:	0.000	490
AVG:	4
BUS V:	0.000	492
AVG:	5
BUS V:	0.000	492
AVG:	6
BUS V:	0.000	492
AVG:	7
BUS V:	0.000	491

**Comment by RobTillaart** [comment]
I reread the datasheet and the averaging function works as a low pass filter internally. 
It does not make multiple readings, but averages with the previous reading, so there is no time penalty

So getting the same time is as it should be, you should see less noise when you make hundreds of readings.

Learned something .

Datasheet, section 8.4.1

_The averaging function is illustrated in Figure 23. Operation begins by first measuring the shunt input signal on
channel 1. This value is then subtracted from the previous value that was present in the corresponding data
output register. This difference is then divided by the value programmed by the averaging mode setting (AVG2-0,
Configuration register bits 11-9) and stored in an internal accumulation register. The computed result is then
added to the previously-loaded data output register value, and the resulting value is loaded to the corresponding
data output register. After the update, the next signal to be measured follows the same process. The larger the
value selected for the averaging mode setting, the less impact or influence any new conversion has on the
average value, as shown in Figure 24. This averaging feature functions as a filter to reduce input noise from the
averaged measurement value._

![image](https://github.com/user-attachments/assets/bbe538a9-f3da-4f65-a07e-10f02f59eabd)

**Comment by RobTillaart** [comment]
@joachymus 

Can you run this program. It tests the timing of the bus conversion time. 

```cpp
//
//    FILE: INA3221_test_BCVT.ino
//  AUTHOR: Rob Tillaart
// PURPOSE: test bus conversion time single channel
//     URL: https://github.com/RobTillaart/INA3221_RT


#include "INA3221.h"


INA3221 INA(0x40);
uint32_t start, stop;


void setup()
{
  Serial.begin(115200);
  Serial.println(__FILE__);
  Serial.print("INA3221_LIB_VERSION: ");
  Serial.println(INA3221_LIB_VERSION);
  Serial.println();

  Wire.begin();

  if (!INA.begin() )
  {
    Serial.println("could not connect. Fix and Reboot");
    while(1);
  }
  else
  {
    Serial.print("Found: \t");
    Serial.println(INA.getAddress());
  }

  //  overwrite default shunts.
  INA.setShuntR(0, 0.100);

  //  should give roughly same timing
  //  see section 8.4.1 datasheet.
  for (int bcvt = 0; bcvt < 8; bcvt++)
  {
    INA.setBusVoltageConversionTime(bcvt);
    Serial.print("BVCT:\t");
    Serial.println(INA.getBusVoltageConversionTime());
    delay(100);

    start = micros();
    float busVoltage = INA.getBusVoltage(0);
    stop = micros();
    Serial.print("BUS V:\t");
    Serial.print(busVoltage, 3);
    Serial.print("\t");
    Serial.println(stop - start);
    delay(100);
  }
}


void loop()
{
}


//  -- END OF FILE --
```

**Comment by joachymus** [comment]
... sorry, I had to leave...

BVCT:	0
BUS V:	0.000	492
BVCT:	1
BUS V:	0.000	491
BVCT:	2
BUS V:	0.000	492
BVCT:	3
BUS V:	0.000	490
BVCT:	4
BUS V:	0.000	492
BVCT:	5
BUS V:	0.000	492
BVCT:	6
BUS V:	0.000	492
BVCT:	7
BUS V:	0.000	492

Not sure why there is no voltage reading...

**Comment by RobTillaart** [comment]
mmm, this is strange ...

> ... sorry, I had to leave...

My agenda also dictates other actions, so to be continued later
Thanks for testing,

**Comment by RobTillaart** [comment]
Lunch break,
an extra print statement if the setBCVT is executed .
Please give it a try.

```cpp
//
//    FILE: INA3221_test_BCVT.ino
//  AUTHOR: Rob Tillaart
// PURPOSE: test bus conversion time single channel
//     URL: https://github.com/RobTillaart/INA3221_RT


#include "INA3221.h"


INA3221 INA(0x40);
uint32_t start, stop;


void setup()
{
  Serial.begin(115200);
  Serial.println(__FILE__);
  Serial.print("INA3221_LIB_VERSION: ");
  Serial.println(INA3221_LIB_VERSION);
  Serial.println();

  Wire.begin();

  if (!INA.begin() )
  {
    Serial.println("could not connect. Fix and Reboot");
    while(1);
  }
  else
  {
    Serial.print("Found: \t");
    Serial.println(INA.getAddress());
  }

  Serial.print("CFG:\t");
  Serial.println(INA.getConfiguration(), HEX);

  //  overwrite default shunts.
  INA.setShuntR(0, 0.100);
  INA.enableChannel(0);

  //  should give roughly same timing
  //  see section 8.4.1 datasheet.
  for (int bcvt = 0; bcvt < 8; bcvt++)
  {
    Serial.println(INA.setBusVoltageConversionTime(bcvt));
    Serial.print("BVCT:\t");
    Serial.println(INA.getBusVoltageConversionTime());
    delay(100);

    start = micros();
    float busVoltage = INA.getBusVoltage(0);
    stop = micros();
    Serial.print("BUS V:\t");
    Serial.print(busVoltage, 3);
    Serial.print("\t");
    Serial.println(stop - start);
    delay(100);
  }
}

void loop()
{
}

//  -- END OF FILE --
```

----

update:  explicitly added INA.enableChannel(0); in setup.

**Comment by RobTillaart** [comment]
>    Serial.println(INA.getConfiguration()); - When reading the config I get all params HIGH (0100 1111 1111 1111) 

According to the datasheet the power on reset value should be 7127h ==> 0111 0001 0010 0111

mode = 7
VSHCT = 4
VBCT = 4
AVG = 0
enable = 7 (all channels)
RST = 0.

**Comment by joachymus** [comment]
Concerning 0100 1111 1111 1111 - it's my settings to all params 7 (avg, bcvt, svct); The default is as above

After explicite enabling the channel:
0
BVCT:	0
BUS V:	6.528	492
0
BVCT:	1
BUS V:	6.520	493
0
BVCT:	2
BUS V:	6.528	492
0
BVCT:	3
BUS V:	6.528	493
0
BVCT:	4
BUS V:	6.528	493
0
BVCT:	5
BUS V:	6.528	493
0
BVCT:	6
BUS V:	6.528	493
0
BVCT:	7
BUS V:	6.528	493

**Comment by RobTillaart** [comment]
mmm (thinking sounds)

At least we get a voltage, however the conversion time does not change.
Back to the datasheet 
(later to be continued)

**Comment by joachymus** [comment]
I got the it working... Not sure how :( I replaced the library files and "corrected" them again. I also removed the INA, used a new one, and then returned the old one - and it works. Unfortunately, different way then I expected (as you described above - it averages with the previous values) which makes it more stable - unfortunately too much stable... It "ramps" for seconds with the higher settings so it's not good for "immediate"/fluctuating value readings...

Concerning the "conversion time", it most probably works also difgferent way then I understood (still don't get it). The time needed for the reading is the same everytime. It probably affects only internal processes and spits "actual" values at the instant of request, kind of in the middle of processing - not waiting for finishing the counts/cycles, or whatever.

Anyway, thank you for your great help, library IS working for me - now I need to find the sweet spot in the settings.

In a preparation for my night shift, I am going to bed so I will be unavailable for some time. Will reply later if anything.

Thanks again!
BR,
j.
